feat(environments): show more status data and allow force stop

lorenzo-cavazzi opened the pull request:
This PR change the table layout of the environments table and adds a couple of extra options for non-running servers:

* show a popup containing kubernetes pod information
* allow force-stop

Preview: https://lorenzotest.dev.renku.ch/

How to test: start one or more interactive environments, check the environment pages and try to force stop.
Projects that can't start correctly:
* https://lorenzotest.dev.renku.ch/projects/lorenzo.cavazzi.tech/slow-build
* https://lorenzotest.dev.renku.ch/projects/lorenzo.cavazzi.tech/failing-image

So far, I have not been able to re-produce a pure "error" state (any suggestion is welcome). Maybe the labeling should be tuned: I will try to identity "pending" states that should be labeled as "error".

fix #361, re #499

rokroskar commented:
You can make an image without `jupyterhub` installed and it will fail to start the container. That should give you a failing container (different from failing to pull an image).

lorenzo-cavazzi commented:
I force-psuhed bacause I forgot to include `Notebooks.state` :man_facepalming: . I also updated https://lorenzotest.dev.renku.ch/

@rokroskar thanks, I'll try that.

lorenzo-cavazzi commented:
@rokroskar it seems that the workflow is the same also in that case. The `notebook` pod state hangs in state `waiting`, therefore no "error" state

rokroskar commented:
ah true, I guess for an error it would have to be something like a missing volume or a secret/configmap etc.

ciyer commented:
Found an issue when testing:
- Start an environment on one of the project above (I used slow-build)
- Kill the environment
- Try to start a new environment for the same project -- this is not possible.

rokroskar commented:
This could be an issue with jupyterhub that we've seen before, where the server remains in the db even though it doesn't exist anymore.

lorenzo-cavazzi commented:
I believe this is related to the `force` delete introduced by this PR SwissDataScienceCenter/renku-notebooks#205 . When the delete is forced, only the k8s code is used to remove the pod and probably jupyterhub stase stays dirty. I'll try to fix that

rokroskar commented:
I believe the problem is only during spawning - if the server is already up, jupyterhub happily notices that it's been deleted and reacts accordingly. If we can pin down the behavior a bit better we can open an issue with jupyterhub.

lorenzo-cavazzi commented:
I did a couple of tests and the problem seems to be indeed related to spwaning pods. JupyterHub is sloppy in reacting also to running pods removal, but if eventually does -- if you remove a running pod from rancher and you try to start a new environment immediatly after, it doesn't work.

I guess it worth to update to `JupyterHub 1.0.0` before opening any issue, but I have not been able to succeed in that. I'll try again soon.

lorenzo-cavazzi commented:
I removed the code to force stop an environment (I left it commented out since it will be re-used soon for #414 ). #361 can't be addressed with the current force stop feature since it leaves JupyterHub in a dirty state and the user won't be able to start a new environment from the same branch/commit until the hub is rebooted (and he doesn't get feedback of this failure). Fortunately, updating JupyterHub to 1.0.0 seems to give us another way to cope with the hanging pods -- this will be addressed separately.

I squashed in 2 commits to keep the history clean.
